torch/_dynamo/variables/builtin.py

```diff
@@ -1095,7 +1095,9 @@ def call_getattr(
                             for i, b in enumerate(bases)
                         ]
                     elif len(bases) == 1 and (
-                        bases[0] is object or bases[0] is torch._C.TensorBase
+                        bases[0] is object
+                        or bases[0] is torch._C.TensorBase
+                        or bases[0] is torch.Tensor
```

Comment:
can you add a more direct test for this somewhere (in maybe test/dynamo/test_misc.py?) Brian told me this was necessary for the case where we construct a Tensor subclass and attempt to trace over a function that does tree_flatten on a container holding the subclass.

Reply:
Sure, let me see add a test case there

Reply:
do you think this is specifically related to high order ops so I need to add tests to test_high_order_ops.py or it's a general problem where pytree flatten a tensor subclass?

Reply:
I don't think that this is related to higher order ops - we only ran into issues with the higher order op because this  was causing a graph break inside the body of our higher order op.

So one test I think we can write is just to `torch.compile(full_graph=True)` some code that tries to pytree a tensor subclass:
```
@torch.compile(backend="aot_eager", fullgraph=True)
def f(subclass):
    nested_list = [[subclass]]
    flattened_list, _ = pytree.tree_flatten(nested_list)
    return flattened_list[0]
```

Reply:
I wrote a test case like this but it does not seems fail even when we don't have this changes, weird.

```
    def test_tracing_py_tree_tensor_subclass(self):
        import torch.utils._pytree as pytree
        from torch.testing._internal.two_tensor import TwoTensor

        def fn(xs):
            nested_xs = [[xs]]
            flat_xs, spec = pytree.tree_flatten(xs)
            return flat_xs[0].clone()

        xs = TwoTensor(torch.ones(2, 2), torch.ones(2, 2))

        counter = CompileCounter()
        torch._dynamo.optimize(counter, nopython=True)(fn)(xs)
        self.assertEqual(counter.frame_count, 1)
        self.assertEqual(counter.op_count, 1)
```

Reply:
oh hmm, it looks like this ends up working out because we hit this line, and the input subclass has a source in dynamo: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builtin.py#L1077

Hmm... I wonder how we can trigger the path where our subclass doesn't have a source (which is apparently what gets triggered in the checkpointing case

Reply:
I'm not actually sure how to trigger a situation where dynamo doesn't have a source for a given tensor (maybe Voz/Yanbo/someone from dynamo can help).

But otherwise - do you think you can try adding a test that uses a subclass + activation checkpointing, and see if that can give us a repro?

Reply:
> Hmm... I wonder how we can trigger the path where our subclass doesn't have a source (which is apparently what gets triggered in the checkpointing case

Does it work if you construct a subclass inside of the function being torch.compile'd, then put that into a list, and attempt to tree_flatten it?

Reply:
Hmmm I don't think we support constructing a subclass inside a torch.compiled function, right now it would graph break when those case happened iiuc

Reply:
I'll try out next put subclass + AC together and see if that repros

Reply:
added AC together and it works now :) Thanks for suggesting!